[18.0][MIG]/[ADD] ddmrp_packaging_product_replace

ThiagoMForgeFlow opened the pull request:
Migration of proposed changes in v14.0 that were never merged.
Original PR:
* https://github.com/OCA/ddmrp/pull/193


@ForgeFlow

ThiagoMForgeFlow committed:
[IMP] ddmrp_packaging_product_replace: pre-commit auto fixes

ThiagoMForgeFlow committed:
[MIG] ddmrp_packaging_product_replace: Migration to 18.0

LoisRForgeFlow commented:
@ThiagoMForgeFlow Could you "fixup the fixups"? :laughing:

ThiagoMForgeFlow commented:
> @ThiagoMForgeFlow Could you "fixup the fixups"? 😆

Done!

LoisRForgeFlow commented:
@ThiagoMForgeFlow Could you also split commit per module changed? I see this PR is modifying also `ddmrp_product_replace`.

TDu committed:
[IMP] ddmrp_product_replace: apply extract method refactor

TDu committed:
[ADD] ddmrp_packaging_product_replace

There seems to be a different outcome when doing a product replacement with packaging set on the related stock buffers between v13 and v14. In v13 the following test passes. But on v14 the new stock buffer created has the packaging of the old product set on them and the constraint  raises an error.

ThiagoMForgeFlow commented:
> @ThiagoMForgeFlow Could you also split commit per module changed? I see this PR is modifying also `ddmrp_product_replace`.

Done, Thanks!

LoisRForgeFlow commented:
@ThiagoMForgeFlow Could you keep authorship of @TDu in the 2 first commits as they are based on his v14 PR?

ThiagoMForgeFlow commented:
> @ThiagoMForgeFlow Could you keep authorship of @TDu in the 2 first commits as they are based on his v14 PR?

Authorship added!

twalter-c2c commented on ddmrp_packaging_product_replace/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,7 @@
+packaging of the old product set on them and the constraint
+\_check_product_packaging raises an error.
+
+Strangely, this is not happening with Odoo 13 ?
```
I wonder if this line is still worth keeping in v18.

ThiagoMForgeFlow replied:
You are right, line removed.

twalter-c2c commented on ddmrp_product_replace/wizards/ddmrp_product_replace.py:
```diff
@@ -123,15 +123,18 @@ def _do_replacement_use_existing(self):
 
+    def _get_new_buffer_default_value(self, replaced_buffer):
+        return dict(
+            product_id=self.new_product_id.id,
```
Thanks for improving the readability by moving this part to a separate method. You should use the `replaced_buffer` instead of `self` if I am not mistaken.
```suggestion
            product_id=replaced_buffer.new_product_id.id,
```

ThiagoMForgeFlow replied:
We should use `self` because the field `new_product_id` is from `ddmrp.product.replace` model, not `stock.buffer`. But I have removed the `replaced_buffer` parameter because is not needed in the function.

twalter-c2c replied:
You are right, thank you for the explanation. So yes, the `replaced_buffer` was not needed indeed (and even misleading apparently) in such a case.

twalter-c2c reviewed (commented):
Thanks! Looks good, small remarks to check.

LoisRForgeFlow reviewed (approved):
:+1:

LoisRForgeFlow commented:
/ocabot merge nobump
